### PR TITLE
[Group 7] Enable nullable annotations for `Microsoft.Extensions.Logging.Console`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to https://github.com/dotnet/runtime/issues/43605

One more after this and we are done 🥳

Questions:
- Both `Formatter` and `Options` can be `null` in `ConsoleLogger.Log()`. `ConsoleLogger` is internal, and they are always set, but there's nothing that guarantees it.
- `console` can be null here. And again, this is internal, so it won't be `null`, but still.
https://github.com/dotnet/runtime/blob/063d08018602291f8dbd9d9b6055b8e1af874cb8/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs#L57

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs [hunk 1]
```diff
@@ -53,8 +56,8 @@ public virtual void EnqueueMessage(LogMessageEntry message)
         // for testing
         internal virtual void WriteMessage(LogMessageEntry entry)
         {
-            IConsole console = entry.LogAsError ? ErrorConsole : Console;
-            console.Write(entry.Message);
+            IConsole? console = entry.LogAsError ? ErrorConsole : Console;
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:59]: seems like ErrorConsole and Console are getting set in the ConsoleLoggerProvider right after ConsoleLoggerProcessor gets instantiated. Maybe in a next PR we could update the ConsoleLoggerProcessor itself to set ErrorConsole and Console so this `console` assignment becomes improved and you wouldn't need `?` or `!`

**Review by maryamariyan — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs [hunk 2]
```diff
@@ -40,7 +43,7 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
             }
             t_stringWriter ??= new StringWriter();
             LogEntry<TState> logEntry = new LogEntry<TState>(logLevel, _name, eventId, state, exception, formatter);
-            Formatter.Write(in logEntry, ScopeProvider, t_stringWriter);
+            Formatter!.Write(in logEntry, ScopeProvider, t_stringWriter);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs:46]: Since this type is internal, can we just change the ctor to take the `ConsoleFormatter` and make the `Formatter` property non-nullable?

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs [hunk 3]
```diff
@@ -53,14 +56,14 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
             {
                 sb.Capacity = 1024;
             }
-            _queueProcessor.EnqueueMessage(new LogMessageEntry(computedAnsiString, logAsError: logLevel >= Options.LogToStandardErrorThreshold));
+            _queueProcessor.EnqueueMessage(new LogMessageEntry(computedAnsiString, logAsError: logLevel >= Options!.LogToStandardErrorThreshold));
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs:59]: Same question for `Options` as `Formatter` above.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/SystemdConsoleFormatter.cs [hunk 4]
```diff
@@ -42,8 +44,8 @@ public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeP
             LogLevel logLevel = logEntry.LogLevel;
             string category = logEntry.Category;
             int eventId = logEntry.EventId.Id;
-            Exception exception = logEntry.Exception;
-            // systemd reads messages from standard out line-by-line in a '<pri>message' format.
+            Exception? exception = logEntry.Exception;
+            // system reads messages from standard out line-by-line in a '<pri>message' format.
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/SystemdConsoleFormatter.cs:48]: This should be reverted.

**Comment by eerhardt** [comment]
>         public ConsoleLoggerProvider(Microsoft.Extensions.Options.IOptionsMonitor<Microsoft.Extensions.Logging.Console.ConsoleLoggerOptions> options, System.Collections.Generic.IEnumerable<Microsoft.Extensions.Logging.Console.ConsoleFormatter> formatters) { }

I believe the `formatters` parameter of the `ConsoleLoggerProvider` ctor should be nullable.

---
Refers to: src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs:75 in 9654d1d. [](commit_id = 9654d1d1412ed3b72aa5cc77d29bc02c9c9f1ec9, deletion_comment = False)

File: src/libraries/Microsoft.Extensions.Logging.Console/src/SystemdConsoleFormatter.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/SystemdConsoleFormatter.cs:48]:
```suggestion
            // systemd reads messages from standard out line-by-line in a '<pri>message' format.
```

**Review by maryamariyan — APPROVED** [review]

File: src/libraries/Common/src/Extensions/Logging/NullScope.cs [hunk 5]
```diff
@@ -1,6 +1,8 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+#nullable enable
```
Comment (eerhardt) [inline src/libraries/Common/src/Extensions/Logging/NullScope.cs:4]: Why are these changes necessary for the code in "Common/src"? They are included in other project with nullability enabled already. So why does this change need them?

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs [hunk 6]
```diff
@@ -46,18 +47,18 @@ public ConsoleLoggerProvider(IOptionsMonitor<ConsoleLoggerOptions> options, IEnu
             ReloadLoggerOptions(options.CurrentValue);
             _optionsReloadToken = _options.OnChange(ReloadLoggerOptions);
 
-            _messageQueue = new ConsoleLoggerProcessor();
-
+            IConsole? console, errorConsole = null;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs:50]:
```suggestion
            IConsole? console;
            IConsole? errorConsole;
```

1. The variables are unconditionally set, so there is no reason to initialize them to `null`.
2. Combining variable declarations is harder to read.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs [hunk 7]
```diff
@@ -19,8 +20,10 @@ internal class ConsoleLoggerProcessor : IDisposable
         public IConsole Console;
         public IConsole ErrorConsole;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:21]:
```suggestion
        public IConsole Console { get; }
        public IConsole ErrorConsole { get; }
```

These don't need to be settable anymore.

**Review by eerhardt — COMMENTED** [review]
Just a couple last comments. After these are addressed I think this will be ready to merge.

File: src/libraries/Common/src/Extensions/Logging/NullScope.cs (hunk 5 above)
Comment (maxkoshevoi) [reply src/libraries/Common/src/Extensions/Logging/NullScope.cs:4]: They are included here as links, and I didn't know that nullability is enabled for them in other project.
Since there're no nullable annotations in them, it should be ok to revert those changes (will do this in a couple of hours).

Comment (eerhardt) [reply src/libraries/Common/src/Extensions/Logging/NullScope.cs:4]: There's critical issue with having these changes, but it does clutter the file unnecessarily. So unless they are required for some reason, let's just revert them.

Comment (maxkoshevoi) [reply src/libraries/Common/src/Extensions/Logging/NullScope.cs]: Done. Also removed it from `NullExternalScopeProvider`. From what I've understood, it's not needed anymore

**Review by eerhardt — APPROVED** [review]
Looks great. Thanks for the great work here @maxkoshevoi.

**Comment by maryamariyan** [comment]
@maxkoshevoi seems like there's a merge conflict on this PR need to be resolved before we could merge.

**Comment by eerhardt** [comment]
Failure is https://github.com/dotnet/runtime/issues/67712.
Merging.
